citrination_client/views/data_view_builder.py

```diff
@@ -58,6 +61,69 @@ def add_raw_descriptor(self, descriptor):
         """
         self.configuration['descriptors'].append(descriptor)
 
+    def add_relation(self, inputs, output, relation_type='lolo'):
+        """
+        Add a manual relation.  If no relations are manually added, Citrination will automatically generate
+        relations when the view is created.
+
+        :param inputs: Array of strings or a single string of descriptor key(s) for the relation inputs
+        :param output: Single string descriptor key for the relation output
+        :param relation_type: Kind of relation, currently only 'lolo' is supported
+        """
+        relation_obj = {}
+
+        if isinstance(inputs, str):
+            relation_obj['inputs'] = [inputs]
+        elif isinstance(inputs, list):
+            if len(inputs) == 0:
+                raise ValueError("Inputs list must not be empty")
+            relation_obj['inputs'] = inputs
+        else:
+            raise ValueError("Unexpected type for inputs, expecting either a string or list of strings")
+
+        if isinstance(output, str):
+            relation_obj['output'] = output
+        else:
+            raise ValueError("Unexpected type for output, expecting a string")
+
+        if relation_type != 'lolo':
+            raise ValueError("Currently, \'lolo\' is the only allowed relation type")
+
+        relation_obj['type'] = relation_type
+
+        # check for duplicate
+        existing_relations = self.configuration['relations']
+        for pos, ele in enumerate(existing_relations):
+            if existing_relations[pos]['inputs'] == relation_obj['inputs'] and \
+                    existing_relations[pos]['output'] == relation_obj['output']:
+                raise ValueError("This relation duplicates an existing relation")
+
+        # check limits
+        if len(existing_relations) > MAX_USER_RELATIONS:
+            raise ValueError("Maximum Relations Reached: Citrination only supports " + str(MAX_USER_RELATIONS) +
+                             " user-defined relations. Please review the existing relations")
+
+        # check inputs exist
+        for input_index, input in enumerate(relation_obj['inputs']):
+            found = False
+            for desc_index, desc in enumerate(self.configuration['descriptors']):
```

Comment:
I think we check the validity of a descriptor key enough times that it's worth creating a `set` of the descriptor keys before this loop begins. Then we can avoid the nested loop and `found` variable, it would just become `if desc in descriptor_keys_set`

Reply:
Set works great for this.  Updated on latest commit.